### PR TITLE
feat(fennel): update parser and queries

### PR DESCRIPTION
**Pull request by alexmozaidze**
WIP version of the parser. It works fine, but there are some bugs.

If you think I should do something that's not in TODO, please let me know ^^

## TODO

- [ ] Correctly distinguish function docstring/metadata from function body (function body's first string/table gets parsed as docstring/metadata)
- [x] Detect double-quoted special table metadata keys (`:fnl/docstring` is detected, while `"fnl/docstring"` is parsed as a regular string)
- [x] Separate `;` from the comment body
- [ ] Support `comment` form (if possible)
- [ ] Ensure form error recovery doesn't leak outside its own paren scope
- [ ] Refactor internals

## Queries

I used `#has-ancestor?`, which is available in Neovim >0.10. Removing it would only make highlighting a *little* incorrect, which would be fine.
I also Removed queries I couldn't use and test, so that there are no broken queries.

## Notes on forms support

I only support forms that have destructuring, complex syntax, or forms that might need to be queried with `#has-ancestor?`/`#has-parent?`.

### CODE REVIEW

**Comment by clason** [comment]
Why did you delete fold and local queries?

And, yes, avoiding `#has-ancestor?` would be good.

File: lua/nvim-treesitter/parsers.lua [hunk 1]
```diff
@@ -531,7 +531,8 @@ list.faust = {
 list.fennel = {
   install_info = {
     url = "https://github.com/alexmozaidze/tree-sitter-fennel",
-    files = { "src/parser.c" },
+    files = { "src/parser.c", "src/scanner.c" },
+    branch = "dev",
```
Comment (clason) [inline lua/nvim-treesitter/parsers.lua:535]: I don't think we want a prerelease branch here; it's OK to leave the PR up for testing, but whatever gets merged must refer to the default branch.

Comment (alexmozaidze) [reply lua/nvim-treesitter/parsers.lua:535]: This is a draft. I will either force-push or make a separate PR for when the above TODO list is mostly done, and there are no major bugs.

**Comment by alexmozaidze** [comment]
I couldn't configure Neovim to use tree-sitter folds, so I can't test those. I will try to add fold queries once I configure tree-sitter folds.

As for locals: I don't know much things that make use of it, so I don't think it's viable to support it. If it's absolutely necessary to support them, I will re-add the queries back, but I won't be able to guarantee that they will be correct.

**Comment by clason** [comment]
> I couldn't configure Neovim to use tree-sitter folds, so I can't test those. I will try to add fold queries once I configure tree-sitter folds.

They're simple enough to easily adapt to changes; removing them is an unacceptable regression.

> As for locals: I don't know much things that make use of it, so I don't think it's viable to support it. If it's absolutely necessary to support them, I will re-add the queries back, but I won't be able to guarantee that they will be 100% correct.

They're used in `incremental_selection` and in downstream plugins; most notably `nvim-treesitter-refactor`. Again, better to leave them in, as long as tests pass.

**Comment by alexmozaidze** [comment]
I seem to have fixed all the major issues.

I'm not sure if I can implement the comment form nicely, since I don't think Tree-sitter can reduce a sequence of expressions into a single one, so making it work won't be pretty. I will most likely resort to `#offset!` since it's the easiest solution I can think of.

Sorry for the radio silence, I recently got myself a job and was pretty busy. I'll refactor the parser and open a new PR in the near future with everything sorted out.

**Comment by clason** [comment]
Please rebase on latest `master`.

File: queries/fennel/highlights.scm [hunk 2]
```diff
@@ -12,44 +12,99 @@
...
-  (symbol) @function.call)
+  call: (symbol) @function.call)
 
 (list
-  .
-  (multi_symbol
-    member: (symbol_fragment) @function.call .))
+  call:
+    (multi_symbol
+      member: (symbol_fragment) @function.call .))
```
Comment (lucario387) [inline queries/fennel/highlights.scm:54]: After you rebase, remember to re-run the formatter as well

**Comment by alexmozaidze** [comment]
I was mistaken when I said that I fixed *all* the major issues. The issue in the first todo bullet point is only partly fixed. The docstring works fine after I applied some patches, but the metadata table is not fixed; it parses a regular table as metadata if it's the only thing in function's body.

Anyhow, this version of the parser works well enough for most use-cases. I'll be fixing problems as I go.

**Comment by alexmozaidze** [comment]
I'll move this to a new PR.

**Comment by alexmozaidze** [comment]
Moved to #6460

**Pull request closed without merging**